kythe/go/util/riegeli/writer.go

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package riegeli
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+)
+
+func (w *Writer) ensureFileHeader() error {
+	if w.fileHeaderWritten {
+		return nil
+	}
+
+	// https://github.com/google/riegeli/blob/master/doc/riegeli_records_file_format.md#file-signature
+	_, err := (&chunk{Header: chunkHeader{ChunkType: fileSignatureChunkType}}).WriteTo(w.w, w.w.pos)
+	// TODO(schroederc): encode RecordsMetadata chunk
+	return err
+}
+
+func (w *Writer) flushRecord() error {
+	if w.record == nil || w.record.numRecords == 0 {
+		// Skip writing empty record chunk.
+		return nil
+	}
+
+	data := w.record.encode()
+	chunk := &chunk{
+		Header: chunkHeader{
+			ChunkType:       recordChunkType,
+			DataSize:        uint64(len(data)),
+			DecodedDataSize: w.record.decodedSize,
+			NumRecords:      w.record.numRecords,
+		},
+		Data: data,
+	}
+	_, err := chunk.WriteTo(w.w, w.w.pos)
+	w.record = newRecordChunk(w.opts)
+	return err
+}
+
+// A blockWriter interleaves blockHeaders inside chunks of data.  Each
+// blockHeader interrupts a single chunk, providing both its relative starting
+// and ending positions.
+type blockWriter struct {
+	w   io.Writer
+	pos int
+}
+
+// Write implements the io.Writer interface.  The given []byte must be a single
```

Comment:
If this is really meant to be a record write, I think we should consider using a name other than `io.Writer` which doesn't imply any framing. Someone passing this into, say `fmt.Fprint`, would be surprised to discover multiple records if the implementation happens to issue two writes.

Maybe `WriteRecord`?

Reply:
✓